site/app/authentication/PamAuthentication.php

```diff
@@ -18,7 +18,7 @@
  */
 class PamAuthentication extends AbstractAuthentication {
     public function authenticate() {
-        if ($this->user_id === null || $this->password === null ||
+        if ( strlen($this->user_id) === 0 || strlen($this->password) === 0 ||
```

Comment:
Let's change this to `empty($this->user_id) || empty($this->password)` to retain the previous logic of preventing null as well as the additional logic of protecting against empty strings.

Reply:
Thanks for reviewing my PR @MasterOdin . I have changed to `strlen(x) === 0' to `empty(x)` but I will be taking time for adding testcases because I don't know how to add testcases.Can you please guide me to this particular issue I am facing ?

Reply:
I probably think they are already present in https://github.com/Submitty/Submitty/blob/master/site/tests/app/authentication/PamAuthenticationTester.php as following name:-
1) `testNoUserId()`
2) `testNoPasswordId()`

Reply:
Sure, we use [PHPUnit](https://phpunit.de/) to handle testing the PHP code and that lives under site/tests with the file I linked above (PamAuthenticationTester) handling unit testing of this authentication method. Generally, each Tester class will either extend from the TestCase class of PHPUnit or our custom BaseUnitTest class that contains a number of helper functions for us to create mocks and things that we use in a bunch of tests and would be cumbersome to copy/paste all around the test suite.

For this particular issue, the best thing to base your work off would be:
https://github.com/Submitty/Submitty/blob/fa90f1e56a37b8ae799458726601d185a48e96b0/site/tests/app/authentication/PamAuthenticationTester.php#L66-L72
where you would use that function to create two new ones:
- testEmptyUsername
- testEmptyPassword

where the first test would be just replacing the call to `setUserId` with an empty string and the second test would be adding a new line that would be `$pam->setPassword('');` while having something non-empty passed to `setUserId`.

Line by line explanation of what the test above does is:
1. create a [mock / test-double](https://phpunit.de/manual/6.5/en/test-doubles.html) for our Core class (though it's not really used for this test)
2. Instantiate an instance of the PamAuthentication class which is what we're testing
3. Set the userId for the PamAuthentication instance
4. Assert that the return of calling the `authenticate` function is false.

https://phpunit.de/manual/6.5/en/appendixes.assertions.html is a good place to look through for all of the different types of asserts you can use in your code. For this, assertFalse is fine.

Reply:
The existing `testNoUserId` and `testNoPassword` test functions only validate correct behavior if the values are null for user_id and password respectively. I'd like to add testcases for if we get an empty string since this was a new bug and to prevent any sort of regressions going forward.

Reply:
Thanks for helping me @MasterOdin I will take some time to learn `PHPUnit` and I will commit all left changes soon.

Reply:
@MasterOdin please review at my PR. I have added two `PHPUnit` testing;